src/Cake.Core/IO/NuGet/AssemblyFilePathFrameworkNameParser.cs

```diff
@@ -0,0 +1,75 @@
+using System;
+using System.Linq;
+using System.Runtime.Versioning;
+
+namespace Cake.Core.IO.NuGet
+{
+    /// <summary>
+    /// Represents an object that parses the segments of a filepath for a .Net framework name
+    /// </summary>
+    public class AssemblyFilePathFrameworkNameParser : IAssemblyFilePathFrameworkNameParser
+    {
+        private static readonly FrameworkName _unsupportedFrameworkName = new FrameworkName("Unsupported", new Version());
+        private readonly IFrameworkNameParser _frameworkNameParser;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="AssemblyFilePathFrameworkNameParser"/> class.
+        /// </summary>
+        /// <param name="frameworkNameParser">The framework name parser.</param>
+        public AssemblyFilePathFrameworkNameParser(IFrameworkNameParser frameworkNameParser)
+        {
+            _frameworkNameParser = frameworkNameParser;
+        }
+        
+        /// <summary>
+        /// Parses the framework name from assembly file path.
+        /// </summary>
+        /// <param name="path">The assembly file path.</param>
+        /// <returns>the parsed framework name -- null when path contains no folders</returns>
+        public FrameworkName Parse(FilePath path)
+        {
+            if (path == null)
+            {
+                throw new ArgumentNullException("path");
+            }
+
+            // The path for a reference might look like this for assembly foo.dll:            
+            // foo.dll
+            // sub\foo.dll
+            // {FrameworkName}{Version}\foo.dll
+            // {FrameworkName}{Version}\sub1\foo.dll
+            // {FrameworkName}{Version}\sub1\sub2\foo.dll
+            return ParseFromDirectoryPath(path.GetDirectory());
+        }
+
+        private FrameworkName ParseFromDirectoryPath(DirectoryPath path)
```

Comment:
Not sure exactly what you mean here.  My previous iteration of this interface actually had `ParseFromDirectoryPath` as the lone method, but I thought it was more intuitive (conceptually, and from the consuming perspective) to use the assemblies' `FilePath` to derive the `FrameworkName` instead.  That's obviously just an opinion though :wink: